In [83]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import f1_score, precision_score, recall_score, accuracy_score
from scipy import stats



In [114]:
titanic=pd.read_csv('Titanic-Dataset.csv')
titanic.head(4)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S

In [33]:
titanic.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [34]:
titanic.Age.median()
titanic.loc[titanic.Age.isna(), 'Age']=titanic.Age.median()

In [35]:
titanic.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [36]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [37]:
titanic.drop(['PassengerId', 'Name','Ticket','Cabin'], axis=1,inplace=True)

In [43]:
titanic.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0       3    0  22.0      1      0   7.2500         0
1         1       1    1  38.0      1      0  71.2833         1
2         1       3    1  26.0      0      0   7.9250         0
3         1       1    1  35.0      1      0  53.1000         0
4         0       3    0  35.0      0      0   8.0500         0

In [39]:
titanic['Embarked'].value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

In [40]:
titanic.loc[titanic.Sex=='male','Sex']=0
titanic.loc[titanic.Sex=='female', 'Sex']=1 


In [42]:
titanic.loc[titanic.Embarked=='S','Embarked']=0
titanic.loc[titanic.Embarked=='C','Embarked']=1
titanic.loc[titanic.Embarked=='Q','Embarked']=2
titanic.fillna(0,inplace=True)

C:\Users\Valentin\AppData\Local\Temp\ipykernel_4896\345484016.py:4: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  titanic.fillna(0,inplace=True)


In [45]:
X_train, X_test, y_train, y_test = train_test_split(titanic.drop('Survived',axis=1), titanic.Survived, test_size=0.25, random_state=42)

In [46]:
model = DecisionTreeClassifier()

In [48]:
model.fit(X_train,y_train)

DecisionTreeClassifier()

In [51]:
print('f1', f1_score(y_test,model.predict(X_test)))
print('precision_score', precision_score(y_test,model.predict(X_test)))
print('recall_score', recall_score(y_test,model.predict(X_test)))
print('accuracy_score', accuracy_score(y_test,model.predict(X_test)))

f1 0.6971428571428572
precision_score 0.7093023255813954
recall_score 0.6853932584269663
accuracy_score 0.7623318385650224


### median
* f1 0.67816091954023
* precision_score 0.6941176470588235
* recall_score 0.6629213483146067
* accuracy_score 0.7488789237668162

In [115]:
titanic=pd.read_csv('Titanic-Dataset.csv')
titanic.head(4)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S

In [116]:
titanic.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [98]:
titanic.groupby(['Sex','Pclass'])['Age'].median()

Sex     Pclass
female  1         35.0
        2         28.0
        3         21.5
male    1         40.0
        2         30.0
        3         25.0
Name: Age, dtype: float64

In [99]:
titanic.drop(['PassengerId', 'Name','Ticket','Cabin'], axis=1,inplace=True)

In [101]:
titanic

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    male  22.0      1      0   7.2500        S
1           1       1  female  38.0      1      0  71.2833        C
2           1       3  female  26.0      0      0   7.9250        S
3           1       1  female  35.0      1      0  53.1000        S
4           0       3    male  35.0      0      0   8.0500        S
..        ...     ...     ...   ...    ...    ...      ...      ...
886         0       2    male  27.0      0      0  13.0000        S
887         1       1  female  19.0      0      0  30.0000        S
888         0       3  female   NaN      1      2  23.4500        S
889         1       1    male  26.0      0      0  30.0000        C
890         0       3    male  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

In [102]:
titanic.isna().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

In [103]:
def impute_age(sex, pclass):
    # Определяем средние значения возраста для мужчин и женщин каждого класса
    if sex == 'male':
        age_values = {'1': 40.0, '2': 30.0, '3': 25.0}
    else:
        age_values = {'1': 35.0, '2': 28.0, '3': 21.5}
    
    return age_values[str(pclass)]


# Заполнение пустот в столбце Age
titanic['Age'].fillna(titanic.apply(lambda x: impute_age(x['Sex'], x['Pclass']), axis=1), inplace=True)

C:\Users\Valentin\AppData\Local\Temp\ipykernel_4896\2837217414.py:12: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  titanic['Age'].fillna(titanic.apply(lambda x: impute_age(x['Sex'], x['Pclass']), axis=1), inplace=True)


In [105]:
titanic

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    male  22.0      1      0   7.2500        S
1           1       1  female  38.0      1      0  71.2833        C
2           1       3  female  26.0      0      0   7.9250        S
3           1       1  female  35.0      1      0  53.1000        S
4           0       3    male  35.0      0      0   8.0500        S
..        ...     ...     ...   ...    ...    ...      ...      ...
886         0       2    male  27.0      0      0  13.0000        S
887         1       1  female  19.0      0      0  30.0000        S
888         0       3  female  21.5      1      2  23.4500        S
889         1       1    male  26.0      0      0  30.0000        C
890         0       3    male  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

In [106]:
titanic.isna().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    2
dtype: int64

In [107]:
titanic.loc[titanic.Sex=='male','Sex']=0
titanic.loc[titanic.Sex=='female', 'Sex']=1 

In [108]:
titanic.loc[titanic.Embarked=='S','Embarked']=0
titanic.loc[titanic.Embarked=='C','Embarked']=1
titanic.loc[titanic.Embarked=='Q','Embarked']=2
titanic.fillna(0,inplace=True)

C:\Users\Valentin\AppData\Local\Temp\ipykernel_4896\345484016.py:4: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  titanic.fillna(0,inplace=True)


In [113]:
titanic.head(4)

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0       3    0  22.0      1      0   7.2500         0
1         1       1    1  38.0      1      0  71.2833         1
2         1       3    1  26.0      0      0   7.9250         0
3         1       1    1  35.0      1      0  53.1000         0

In [110]:
X_train, X_test, y_train, y_test = train_test_split(titanic.drop('Survived',axis=1), titanic.Survived, test_size=0.25, random_state=42)

In [111]:
model = DecisionTreeClassifier()
model.fit(X_train,y_train)

DecisionTreeClassifier()

In [118]:
print('f1', f1_score(y_test,model.predict(X_test)))
print('precision_score', precision_score(y_test,model.predict(X_test)))
print('recall_score', recall_score(y_test,model.predict(X_test)))
print('accuracy_score', accuracy_score(y_test,model.predict(X_test)))

f1 0.6779661016949152
precision_score 0.6818181818181818
recall_score 0.6741573033707865
accuracy_score 0.7443946188340808


### median
* f1 0.67816091954023
* precision_score 0.6941176470588235
* recall_score 0.6629213483146067
* accuracy_score 0.7488789237668162

### Замена пустот в Age исходя из параметров по столбцам Sex and Pclass

* f1 0.6779661016949152
* precision_score 0.6941176470588235
* recall_score 0.6629213483146067
* accuracy_score 0.7488789237668162

### Получается, что параметр с использованием median F1=67.8%, а при замене пустот F1=67.79% -метрикии совпадают и в результате замены пустот метрики не улучшились.